### ISSUE TITLE
Propose iter_exchange function

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#247]
```c++
// Who says C++ is verbose?
template <class X, class Y>
concept bool _IndirectlyExchangeable =
  Readable<X>() && Constructible<value_type_t<X>, rvalue_reference_t<X>>() &&
  IndirectlyMovable<Y, X>() && MoveConstructible<value_type_t<X>>();

template <class X, class Y>
constexpr bool __is_nothrow_indirectly_exchangeable_v =
  is_nothrow_constructible_v<value_type_t<X>, rvalue_reference_t<X>> &&
  is_nothrow_assignable_v<reference_t<X>, rvalue_reference_t<Y>> &&
  is_nothrow_move_constructible_v<value_type_t<X>>;

template <class X, class Y>
  requires _IndirectlyExchangeable<remove_reference_t<Y>, remove_reference_t<X>>
constexpr auto iter_exchange(X&& x, Y&& y) 
    noexcept(__is_nothrow_indirectly_exchangeable_v<remove_reference_t<X>, remove_reference_t<Y>>)
{
  value_type_t<remove_reference_t<X>> old_value = iter_move((X&&)x);
  *(X&&)x = iter_move((Y&&)y);
  return old_value;
}
```

**Comment by CaseyCarter** [comment]
Missing `Readable<remove_reference_t<X>>()`.

**Comment by CaseyCarter** [comment]
and I *think* ` Movable<value_type_t<remove_reference_t<X>>>()` can be relaxed to `MoveConstructible`.

**Comment by ericniebler** [comment]
Updated, thanks.

**Comment by ericniebler** [comment]
We need `requires(auto)` and `noexcept(auto)`.

**Comment by CaseyCarter** [comment]
Slightly more legible with some aliases for `remove_reference_t<X/Y>`:

```c++
// Who says C++ is verbose?
template <class X, class Y,
  class Left = remove_reference_t<X>,
  class Right = remove_reference_t<Y>>
  requires Readable<Left>() && 
    Constructible<value_type_t<Left>, rvalue_reference_t<Left>>() &&
    IndirectlyMovable<Right, Left>() &&
    MoveConstructible<value_type_t<Left>>()
constexpr auto iter_exchange(X&& x, Y&& y) 
    noexcept(
        is_nothrow_constructible_v<value_type_t<Left>, 
          rvalue_reference_t<Left>> &&
        is_nothrow_assignable_v<reference_t<Left>,
                                rvalue_reference_t<Right>> &&
        is_nothrow_move_constructible_v<value_type_t<Left>>)
{
  value_type_t<Left> old_value = iter_move((X&&)x);
  *(X&&)x = iter_move((Y&&)y);
  return old_value;
}```

**Comment by CaseyCarter** [comment]
> `// Who says C++ is verbose?`

Three lines of code with **fourteen** lines of declaration. That's something special

**Comment by CaseyCarter** [comment]
This issue is resolved by the PR for #284.
